### PR TITLE
Scheduler: block until all threads are unbound

### PR DESCRIPTION
**Pull request by ben-clayton**
Previously we'd simply assert when there was a thread still bound.
This isn't particularly friendly for multi-threaded logic that needs to
synchronize the unbind() calls before the destructor. Worse still, this
only tells you something is wrong in debug builds.

This issue was uncovered by ensuring the scheduler is bound on the new
thread spun up by `marl::blocking_call()`.
Without forwarding the scheduler to the new thread, attempts to schedule
tasks from the `blocking_call()` would assert, as the scheduler was not
carried forward.

Also: Don't `#include "defer.h"` from the public header - this will
`#define defer()`, which some people may not want.

### CODE REVIEW

File: include/marl/blockingcall.h [hunk 1]
```diff
@@ -32,10 +32,17 @@ class OnNewThread {
...
+    auto scheduler = Scheduler::get();
     auto thread = std::thread(
-        [&](Args&&... args) {
-          defer(wg.done());
+        [&, wg](Args&&... args) {
+          if (scheduler != nullptr) {
+            scheduler->bind();
+          }
           result = f(std::forward<Args>(args)...);
+          wg.done();
```
Comment (ben-clayton) [inline include/marl/blockingcall.h]: TODO: This should probably happen after the call to `unbind` to ensure that all scheduled tasks are completed before the `blocking_call()` returns.

Comment (ben-clayton) [reply include/marl/blockingcall.h]: Done, uncovering more issues than I was expecting (now fixed by this PR).

File: include/marl/blockingcall.h [hunk 2]
```diff
@@ -50,10 +57,17 @@ class OnNewThread<void> {
   template <typename F, typename... Args>
   inline static void call(F&& f, Args&&... args) {
     WaitGroup wg(1);
+    auto scheduler = Scheduler::get();
     auto thread = std::thread(
-        [&](Args&&... args) {
-          defer(wg.done());
+        [&, wg](Args&&... args) {
```
Comment (amaiorano) [inline include/marl/blockingcall.h]: Do you mean to capture 'wg' by copy here?

File: src/scheduler.cpp [hunk 3]
```diff
@@ -99,29 +99,31 @@ void Scheduler::bind() {
...
-    MARL_ASSERT(it != bound->singleThreadedWorkers.end(),
+    auto it = bound->singleThreadedWorkers.byTid.find(tid);
+    MARL_ASSERT(it != bound->singleThreadedWorkers.byTid.end(),
                 "singleThreadedWorker not found");
-    worker = std::move(it->second);
-    bound->singleThreadedWorkers.erase(tid);
+    MARL_ASSERT(it->second.get() == worker, "worker is not bound?");
+    bound->singleThreadedWorkers.byTid.erase(it);
+    if (bound->singleThreadedWorkers.byTid.size() == 0) {
+      bound->singleThreadedWorkers.unbind.notify_one();
```
Comment (amaiorano) [inline src/scheduler.cpp]: Is there only one thread waiting in ~Scheduler, hence the notify_one? Just wondering if it should be notify_all.

Comment (ben-clayton) [reply src/scheduler.cpp]: Correct. The scheduler can only be destructed by one thread (which is the only place that waits for `unbind`), so there's no need for notify_all.

File: include/marl/blockingcall.h (hunk 2 above)
Comment (ben-clayton) [reply include/marl/blockingcall.h]: Yes. All marl synchronisation primitives are copy-by-value (they hold a shared_ptr to the underlying data). This avoids issues where the primitive may be accessed after the stack has been unwound.

**Pull request merged**